### PR TITLE
Implement RFC#1559: allow all literals in attributes

### PR DESCRIPTION
**Pull request by SergioBenitez**
Implemented rust-lang/rfcs#1559, tracked by #34981.

### CODE REVIEW

File: src/librustc/hir/check_attr.rs [hunk 1]
```diff
@@ -53,33 +53,36 @@ impl<'a> CheckAttrVisitor<'a> {
...
-            let word: &str = &word.name();
-            let message = match word {
-                "C" => {
-                    if target != Target::Struct && target != Target::Enum {
-                            "attribute should be applied to struct or enum"
-                    } else {
-                        continue
+            let name = word.name();
+            let message = match name {
+                Some(word) => match &*word {
```
Comment (eddyb) [inline src/librustc/hir/check_attr.rs]: This could probably be `if let Some(word) = word.name() {`.

File: src/librustc_metadata/common.rs [hunk 2]
```diff
@@ -50,21 +50,8 @@ pub const tag_def_key: usize = 0x2c;
...
-pub const tag_meta_item_value: usize = 0x31;
-
 pub const tag_attributes: usize = 0x101; // top-level only
 
-pub const tag_attribute: usize = 0x32;
-
-pub const tag_meta_item_word: usize = 0x33;
-
-pub const tag_meta_item_list: usize = 0x34;
-
```
Comment (eddyb) [inline src/librustc_metadata/common.rs]: These should use the `GAP` comments.

File: src/librustc/hir/check_attr.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc/hir/check_attr.rs]: Wait, doesn't `.name()` cause an error for `repr(5)` currently, that would be silenced?

File: src/libsyntax/ast.rs [hunk 3]
```diff
@@ -1105,6 +1123,26 @@ impl LitKind {
...
+    pub fn is_unsuffixed(&self) -> bool {
+        match *self {
+            LitKind::Str(..) => true,
+            LitKind::ByteStr(..) => true,
+            LitKind::Byte(..) => true,
+            LitKind::Char(..) => true,
+            LitKind::Int(_, LitIntType::Unsuffixed) => true,
+            LitKind::FloatUnsuffixed(..) => true,
+            LitKind::Bool(..) => true,
+            _ => false,
```
Comment (eddyb) [inline src/libsyntax/ast.rs]: Maybe all the cases should be written out, to avoid problems in the future.

File: src/libsyntax/parse/attr.rs [hunk 4]
```diff
@@ -193,9 +193,26 @@ impl<'a> Parser<'a> {
...
 
-    /// matches meta_item = IDENT
-    /// | IDENT = lit
-    /// | IDENT meta_seq
+    fn parse_unsuffixed_lit(&mut self) -> PResult<'a, ast::Lit> {
+        let lit = self.parse_lit()?;
+        debug!("Checking if {:?} is unusuffixed.", lit);
+
+        if !lit.node.is_unsuffixed() {
+            let msg = "suffixed literals are not allowed in meta-items";
```
Comment (eddyb) [inline src/libsyntax/parse/attr.rs]: I don't think we should leak "meta-items" to the user. "in attributes" still seems correct to me and pretty clear.

**Comment by eddyb** [comment]
Looks pretty good to me, great job! :tada:

cc @nrc @cgswords @Manishearth

File: src/librustc_incremental/assert_dep_graph.rs [hunk 5]
```diff
@@ -116,31 +116,38 @@ impl<'a, 'tcx> IfThisChanged<'a, 'tcx> {
         for attr in self.tcx.get_attrs(def_id).iter() {
             if attr.check_name(IF_THIS_CHANGED) {
                 let mut id = None;
-                for meta_item in attr.meta_item_list().unwrap_or_default() {
-                    if meta_item.is_word() && id.is_none() {
-                        id = Some(meta_item.name().clone());
-                    } else {
-                        // FIXME better-encapsulate meta_item (don't directly access `node`)
```
Comment (Mark-Simulacrum) [inline src/librustc_incremental/assert_dep_graph.rs]: Node is still directly accessed in the span_bug! so perhaps the FIXME shouldn't be removed?

File: src/librustc_incremental/assert_dep_graph.rs [hunk 6]
```diff
@@ -116,31 +116,38 @@ impl<'a, 'tcx> IfThisChanged<'a, 'tcx> {
...
             } else if attr.check_name(THEN_THIS_WOULD_NEED) {
                 let mut dep_node_interned = None;
                 let mut id = None;
-                for meta_item in attr.meta_item_list().unwrap_or_default() {
-                    if meta_item.is_word() && dep_node_interned.is_none() {
-                        dep_node_interned = Some(meta_item.name().clone());
-                    } else if meta_item.is_word() && id.is_none() {
-                        id = Some(meta_item.name().clone());
-                    } else {
-                        // FIXME better-encapsulate meta_item (don't directly access `node`)
```
Comment (Mark-Simulacrum) [inline src/librustc_incremental/assert_dep_graph.rs]: Same as above; FIXME might still need to be kept.

File: src/librustc/hir/check_attr.rs (hunk 1 above)
Comment (SergioBenitez) [inline src/librustc/hir/check_attr.rs]: I'll use `if let ...`. The check you're referring to happens [https://github.com/rust-lang/rust/pull/35850/files#diff-21e1f9f42b7976b50c28979848df2e04R951](here).

File: src/libsyntax/ast.rs (hunk 3 above)
Comment (SergioBenitez) [inline src/libsyntax/ast.rs]: That's a good idea. I'll do that.

File: src/librustc_metadata/common.rs (hunk 2 above)
Comment (SergioBenitez) [inline src/librustc_metadata/common.rs]: Got it. Will add.

File: src/librustc_incremental/assert_dep_graph.rs (hunk 5 above)
Comment (SergioBenitez) [inline src/librustc_incremental/assert_dep_graph.rs]: This happens in a lot of places without a similar comment. Is it more important to place such a comment here than in other places?

File: src/libsyntax/parse/attr.rs (hunk 4 above)
Comment (SergioBenitez) [inline src/libsyntax/parse/attr.rs]: Sounds good to me.

File: src/librustc/hir/check_attr.rs (hunk 1 above)
Comment (SergioBenitez) [inline src/librustc/hir/check_attr.rs]: Actually, it's slightly cleaner to use `match word.name()`, so I'll switch it to that.

Comment (eddyb) [inline src/librustc/hir/check_attr.rs]: Might be able to do `match word.name().as_ref().map_or("", |s| s)` (or `map_or("", |s| &**s)`) - assuming that it has been previously validated.

Comment (SergioBenitez) [inline src/librustc/hir/check_attr.rs]: Yeah, I considered that, but we don't know if it's a word at this point (though it doesn't matter too much, yet). Plus, it seems a bit more obvious to specifically match on the None case.

File: src/librustc_incremental/assert_dep_graph.rs (hunk 5 above)
Comment (cgswords) [inline src/librustc_incremental/assert_dep_graph.rs]: I added those fixmes; ultimately, there should be some encapsulation to make meta-items more opaque (so that their internal representation is less obvious and more resilient to changes). If it's done other places, this fixme should be added there, too.

File: src/librustc/hir/check_attr.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc/hir/check_attr.rs]: Two nested `match` are pretty ugly :disappointed:.

File: src/libsyntax/parse/attr.rs [hunk 7]
```diff
@@ -213,36 +230,43 @@ impl<'a> Parser<'a> {
...
-                // FIXME #623 Non-string meta items are not serialized correctly;
-                // just forbid them for now
-                match lit.node {
-                    ast::LitKind::Str(..) => {}
-                    _ => {
-                        self.span_err(lit.span,
-                                      "non-string literals are not allowed in meta-items");
-                    }
-                }
+                let lit = self.parse_unsuffixed_lit()?;
```
Comment (cgswords) [inline src/libsyntax/parse/attr.rs]: Why has this been changed to be an unsuffixed literal? Supporting all literals means supporting suffixed literals (and the parser already has that operation).

File: src/libsyntax/parse/attr.rs [hunk 8]
```diff
@@ -213,36 +230,43 @@ impl<'a> Parser<'a> {
...
                 let hi = self.span.hi;
                 Ok(P(spanned(lo, hi, ast::MetaItemKind::NameValue(name, lit))))
-            }
+            },
```
Comment (cgswords) [inline src/libsyntax/parse/attr.rs]: Nit: we don't need this comma, get rid of it (and same with the one added below)

File: src/libsyntax/ast.rs [hunk 9]
```diff
@@ -456,7 +472,7 @@ pub enum MetaItemKind {
     /// List meta item.
     ///
     /// E.g. `derive(..)` as in `#[derive(..)]`
-    List(InternedString, Vec<P<MetaItem>>),
+    List(InternedString, Vec<MetaListItem>),
```
Comment (cgswords) [inline src/libsyntax/ast.rs]: I am really dubious about removing this pointer. It means we're going to clone a goodly number of literals, which might cause perf regressions in, eg, serde.

File: src/libsyntax/ast.rs (hunk 3 above)
Comment (cgswords) [inline src/libsyntax/ast.rs]: Why add this here, and not to the LitKind impl?

File: src/libsyntax/attr.rs [hunk 10]
```diff
@@ -77,6 +81,93 @@ pub fn is_used(attr: &Attribute) -> bool {
     })
 }
 
+pub trait AttrMetaListItemMethods {
+    /// Returns true if this list item is a MetaItem with a name of `name`.
+    fn check_name(&self, name: &str) -> bool {
+        self.meta_item().map_or(false, |meta_item| meta_item.check_name(name))
```
Comment (cgswords) [inline src/libsyntax/attr.rs]: Nit: You probably want `and_then`.

File: src/libsyntax/attr.rs [hunk 11]
```diff
@@ -77,6 +81,93 @@ pub fn is_used(attr: &Attribute) -> bool {
     })
 }
 
+pub trait AttrMetaListItemMethods {
```
Comment (cgswords) [inline src/libsyntax/attr.rs]: It seems like adding `literal` to `AttrMetaMethods` will be a lot cleaner than defining an entire duplication of `AttrMetaMethods` to add `literal`. It will also make things more uniform in the other parts of the compiler.

File: src/libsyntax/attr.rs [hunk 12]
```diff
@@ -89,8 +180,9 @@ pub trait AttrMetaMethods {
     /// Gets the string value if self is a MetaItemKind::NameValue variant
     /// containing a string, otherwise None.
     fn value_str(&self) -> Option<InternedString>;
+
     /// Gets a list of inner meta items from a list MetaItem type.
-    fn meta_item_list(&self) -> Option<&[P<MetaItem>]>;
+    fn meta_item_list(&self) -> Option<&[MetaListItem]>;
```
Comment (cgswords) [inline src/libsyntax/attr.rs]: Removing the pointer here (+below) may cause perf regressions with serde, etc.

File: src/libsyntax/parse/attr.rs (hunk 7 above)
Comment (SergioBenitez) [inline src/libsyntax/parse/attr.rs]: That was the conclusion we came to in https://github.com/rust-lang/rfcs/pull/1559,

File: src/libsyntax/attr.rs [hunk 13]
```diff
@@ -557,14 +689,19 @@ fn find_stability_generic<'a, I>(diagnostic: &Handler,
...
                     let mut reason = None;
                     for meta in metas {
-                        match &*meta.name() {
-                            "since" => if !get(meta, &mut since) { continue 'outer },
-                            "reason" => if !get(meta, &mut reason) { continue 'outer },
-                            _ => {
-                                handle_errors(diagnostic, meta.span,
-                                              AttrError::UnknownMetaItem(meta.name()));
-                                continue 'outer
+                        if let MetaListItemKind::MetaItem(ref mi) = meta.node {
```
Comment (cgswords) [inline src/libsyntax/attr.rs]: Please write a variant of `list_contains_name` (or change its type signature) to return an `Option`, then rewrite this (and maybe also `cfg_matches` to use that instead of the mess that's here right now.

File: src/libsyntax/ast.rs (hunk 9 above)
Comment (eddyb) [inline src/libsyntax/ast.rs]: `P` is owned though, so they wouldn't be shared, unless you mean just moving the value around? Not only that, but ASTs shouldn't get cloned all the time.

File: src/libsyntax/parse/attr.rs (hunk 7 above)
Comment (cgswords) [inline src/libsyntax/parse/attr.rs]: Well okay, then. :cry:

File: src/libsyntax/ast.rs (hunk 9 above)
Comment (cgswords) [inline src/libsyntax/ast.rs]: Yeah, sorry, that is what I meant (though iirc meta-items do end up getting cloned here and there).

File: src/libsyntax/test.rs [hunk 14]
```diff
@@ -19,8 +19,7 @@ use std::iter;
 use std::slice;
 use std::mem;
 use std::vec;
-use attr::AttrMetaMethods;
-use attr;
+use attr::{self, AttrMetaMethods, AttrMetaListItemMethods};
```
Comment (cgswords) [inline src/libsyntax/test.rs]: Please add some tests explicitly for MetaItemLists here.

File: src/libsyntax/attr.rs [hunk 15]
```diff
@@ -453,26 +573,38 @@ pub fn requests_inline(attrs: &[Attribute]) -> bool {
 
 /// Tests if a cfg-pattern matches the cfg set
 pub fn cfg_matches(cfgs: &[P<MetaItem>], cfg: &ast::MetaItem,
-                   sess: &ParseSess, features: Option<&Features>)
+                   sess: &ParseSess, fs: Option<&Features>)
```
Comment (cgswords) [inline src/libsyntax/attr.rs]: Nit: don't make the variable names worse.

File: src/libsyntax/attr.rs [hunk 16]
```diff
@@ -229,10 +324,14 @@ pub fn mk_name_value_item(name: InternedString, value: ast::Lit)
     mk_spanned_name_value_item(DUMMY_SP, name, value)
 }
 
-pub fn mk_list_item(name: InternedString, items: Vec<P<MetaItem>>) -> P<MetaItem> {
+pub fn mk_list_item(name: InternedString, items: Vec<MetaListItem>) -> P<MetaItem> {
     mk_spanned_list_item(DUMMY_SP, name, items)
 }
 
+pub fn mk_list_word_item(name: InternedString) -> ast::MetaListItem {
+    respan(DUMMY_SP, MetaListItemKind::MetaItem(mk_spanned_word_item(DUMMY_SP, name)))
```
Comment (cgswords) [inline src/libsyntax/attr.rs]: Nit: use `dummy_spanned` instead of `respan` here

File: src/librustc/hir/check_attr.rs (hunk 1 above)
Comment (SergioBenitez) [inline src/librustc/hir/check_attr.rs]: Changed to (from IRC):

```
            let name = match word.name() {
                Some(word) => word,
                None => continue,
            };

            let message = match &*name {
                ...
            }

            span_err!(self.sess, attr.span, E0517, "{}", message);
```

File: src/libsyntax/ast.rs (hunk 3 above)
Comment (SergioBenitez) [inline src/libsyntax/ast.rs]: This is the LitKind impl...

File: src/librustc_driver/lib.rs [hunk 17]
```diff
@@ -666,6 +666,9 @@ impl RustcDefaultCalls {
                             // MetaItemKind::List items in the configuration returned by
                             // `build_configuration`.
                             panic!("MetaItemKind::List encountered in default cfg")
+                        } else {
+                            // There also shouldn't be literals.
+                            panic!("Unexpected MetaItem literal in default cfg!")
```
Comment (cgswords) [inline src/librustc_driver/lib.rs]: Maybe panic about a malformed attribute, not a meta-item.

File: src/libsyntax/parse/attr.rs (hunk 7 above)
Comment (SergioBenitez) [inline src/libsyntax/parse/attr.rs]: Would be a one line change (+ some deletions) to reconsider in the future, however!

**Comment by cgswords** [comment]
There should be some `run-pass` tests.

File: src/libsyntax/ast.rs (hunk 3 above)
Comment (cgswords) [inline src/libsyntax/ast.rs]: Whoops, nevermind. I was thinking of the `Lit` stuff in `token`; disregard!

File: src/libsyntax/ast.rs [hunk 18]
```diff
@@ -1105,6 +1123,26 @@ impl LitKind {
...
+            LitKind::Char(..) => true,
+            LitKind::Int(_, LitIntType::Unsuffixed) => true,
+            LitKind::FloatUnsuffixed(..) => true,
+            LitKind::Bool(..) => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true if this literal has a suffix.
+    pub fn is_suffixed(&self) -> bool {
```
Comment (SergioBenitez) [inline src/libsyntax/ast.rs]: See above.

File: src/libsyntax/attr.rs (hunk 11 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: The issues is that AttrMetaMethod's `name()` always returns an InternedString, while ListMetaMethods returns an Option<...>.

Comment (cgswords) [inline src/libsyntax/attr.rs]: `ListMetaMethods` should probably return the literal as a string for `name()` instead of a `None` (cc @nrc)

File: src/libsyntax/attr.rs (hunk 15 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: This was so line 590/91 would be <= 100 cols.

File: src/librustc_driver/lib.rs (hunk 17 above)
Comment (SergioBenitez) [inline src/librustc_driver/lib.rs]: The panic above mentions meta item, so I thought this would be more uniform.

File: src/libsyntax/attr.rs (hunk 15 above)
Comment (cgswords) [inline src/libsyntax/attr.rs]: I'm pretty sure this doesn't adhere to what `rustfmt` would do, anyway. It should be:

```rust
pub fn cfg_matches(cfgs: &[P<MetaItem>],
                   cfg: &ast::MetaItem,
                   sess: &ParseSess,
                   features: Option<&Features>)
                   -> bool {
```

File: src/librustc_driver/lib.rs (hunk 17 above)
Comment (cgswords) [inline src/librustc_driver/lib.rs]: Change both? It would be nice to give more descriptive errors with some of the values on-hand, too (eg emitting the `config` name that caused the problem).

File: src/libsyntax/attr.rs (hunk 11 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: My opinion is that since this is not the way `name()` is used in the code, it would be error prone to make this change.

**Comment by cgswords** [comment]
Aside from my comments and request for tests, this is basically all awesome. :tada:

File: src/libsyntax/attr.rs (hunk 16 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: Sure thing.

File: src/libsyntax/parse/attr.rs (hunk 8 above)
Comment (SergioBenitez) [inline src/libsyntax/parse/attr.rs]: While we don't need it, I believe this style to be less error prone (if the arm gets changed). It also appears to be the style used in much of the code.

File: src/libsyntax/ast.rs (hunk 9 above)
Comment (SergioBenitez) [inline src/libsyntax/ast.rs]: Can we land this and benchmark?

File: src/libsyntax/attr.rs (hunk 12 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: Same as above.

File: src/libsyntax/attr.rs (hunk 13 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: Oh, that already exists. This just doesn't use it for some reason.

File: src/libsyntax/ast.rs (hunk 9 above)
Comment (cgswords) [inline src/libsyntax/ast.rs]: It's preferable to benchmark before landing, because otherwise it will mean immediately landing a second PR just to fix it. `rustup` will let you use a local build as the default; you can build `serde` and run its perf benchmarks with both.

File: src/libsyntax/attr.rs (hunk 15 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: I'll split the "any" and "all" branches into multiple lines and keep `features`.

File: src/libsyntax/parse/attr.rs (hunk 8 above)
Comment (cgswords) [inline src/libsyntax/parse/attr.rs]: `rustfmt` will remove the commas, and it is generally used to inform the compiler style guide

Comment (SergioBenitez) [inline src/libsyntax/parse/attr.rs]: Sounds good. Will do.

File: src/librustc_incremental/assert_dep_graph.rs (hunk 6 above)
Comment (SergioBenitez) [inline src/librustc_incremental/assert_dep_graph.rs]: Okay, I'll re-add them.

File: src/librustc_driver/lib.rs (hunk 17 above)
Comment (SergioBenitez) [inline src/librustc_driver/lib.rs]: Sounds good.

File: src/libsyntax/ast.rs (hunk 9 above)
Comment (SergioBenitez) [inline src/libsyntax/ast.rs]: Thinking about this: MetaItem almost always sites behind a P<> anyways, so there should be very little, if any, performance degradation. We might even see a performance win once literals start being use due to coalesced allocations.

**Comment by Manishearth** [comment]
When this is ready to merge, let me know, we can batch up everything in https://github.com/rust-lang/rust/issues/31645#issuecomment-239228452

File: src/libsyntax/parse/attr.rs (hunk 7 above)
Comment (nrc) [inline src/libsyntax/parse/attr.rs]: FWIW, I think we should accept suffixed literals, I don't understand the argument against.

File: src/libsyntax/attr.rs (hunk 11 above)
Comment (nrc) [inline src/libsyntax/attr.rs]: Doesn't sounds like there is a good solution here, but at the least something needs renaming or there needs to be some docs here - totally unclear to me what the abstractions here represent

File: src/librustc_borrowck/borrowck/mir/mod.rs [hunk 19]
```diff
@@ -43,8 +43,10 @@ fn has_rustc_mir_with(attrs: &[ast::Attribute], name: &str) -> Option<P<MetaItem
         if attr.check_name("rustc_mir") {
             let items = attr.meta_item_list();
             for item in items.iter().flat_map(|l| l.iter()) {
-                if item.check_name(name) {
-                    return Some(item.clone())
+                if let Some(mi) = item.meta_item() {
+                    if mi.check_name(name) {
+                        return Some(mi.clone())
+                    }
```
Comment (nrc) [inline src/librustc_borrowck/borrowck/mir/mod.rs]: Needing double `if`s here is annoying - could check_name return None for non-meta items? (More generally, I kind of hate the concept of meta-item - it is arbitrary and meaningless. We should do our best not to expose the concept outside of libsyntax).

File: src/librustc_driver/lib.rs [hunk 20]
```diff
@@ -655,17 +655,19 @@ impl RustcDefaultCalls {
...
+                            println!("{}=\"{}\"", cfg.name(), s);
                         } else if cfg.is_meta_item_list() {
                             // Right now there are not and should not be any
                             // MetaItemKind::List items in the configuration returned by
                             // `build_configuration`.
-                            panic!("MetaItemKind::List encountered in default cfg")
+                            panic!("Found an unexpected list in cfg attribute '{}'!", cfg.name())
+                        } else {
+                            // There also shouldn't be literals.
+                            panic!("Found an unexpected literal in cfg attribute '{}'!", cfg.name())
```
Comment (nrc) [inline src/librustc_driver/lib.rs]: Given that this is not meant to be a user-facing error message, I don't think we need two different messages

File: src/librustc_borrowck/borrowck/mir/mod.rs (hunk 19 above)
Comment (SergioBenitez) [inline src/librustc_borrowck/borrowck/mir/mod.rs]: Actually, we can call check_name directly on the list item. I'll change it to that.

File: src/libsyntax/ast.rs [hunk 21]
```diff
@@ -439,6 +439,22 @@ pub struct Crate {
     pub exported_macros: Vec<MacroDef>,
 }
 
+/// A spanned compile-time attribute list item.
+pub type MetaListItem = Spanned<MetaListItemKind>;
```
Comment (nrc) [inline src/libsyntax/ast.rs]: Maybe `NestedMetaItem` - the list part of the name is a bit misleading

**Comment by nrc** [comment]
This should be feature-gated, right?

**Comment by SergioBenitez** [comment]
Why should this be feature gated? The syntax is backwards compatible, and custom attributes are nightly only.

**Comment by eddyb** [comment]
@SergioBenitez So you're saying all stable attributes do checks that prevent use of literals?

**Comment by SergioBenitez** [comment]
@eddyb: That's the intention with this PR, yeah.

**Comment by nrc** [comment]
I would be more paranoid than that - there is nothing to stop someone accidentally accepting a literal in a stable attribute or adding a new attribute which requires a literal.

File: src/test/run-pass-fulldeps/auxiliary/macro_crate_test.rs [hunk 22]
```diff
@@ -42,40 +44,45 @@ pub fn plugin_registrar(reg: &mut Registry) {
...
-                         attr: &MetaItem,
-                         it: Annotatable) -> Vec<Annotatable> {
+                         _sp: Span,
+                         _attr: &MetaItem,
+                         it: Annotatable)
+                         -> Vec<Annotatable> {
     match it {
-        Annotatable::Item(it) => vec![
+        Annotatable::Item(it) => {
+            vec![
```
Comment (eddyb) [inline src/test/run-pass-fulldeps/auxiliary/macro_crate_test.rs]: Why this change? For the indentation to match up you'd need to indent all the lines inside - it's simpler just to keep it the way it is.

Comment (SergioBenitez) [inline src/test/run-pass-fulldeps/auxiliary/macro_crate_test.rs]: This was `rustfmt`. Happy to change it back.

Comment (eddyb) [inline src/test/run-pass-fulldeps/auxiliary/macro_crate_test.rs]: cc @nikomatsakis @ubsan Another case where rustfmt is ugly *and* wrong (nesting without indentation).

Comment (ubsan) [inline src/test/run-pass-fulldeps/auxiliary/macro_crate_test.rs]: Weeeeeeee

Comment (nikomatsakis) [inline src/test/run-pass-fulldeps/auxiliary/macro_crate_test.rs]: @eddyb I don't think this ugly -- it's more "regular", if more "wordy" -- but yes the indentation is a bug. Probably worth filing an issue if we can isolate it.

Comment (ubsan) [inline src/test/run-pass-fulldeps/auxiliary/macro_crate_test.rs]: This would be another case where a real rustfmt style guide would help. What is the correct thing to do here? ping @nrc

Comment (nrc) [inline src/test/run-pass-fulldeps/auxiliary/macro_crate_test.rs]: https://github.com/rust-lang-nursery/rustfmt/issues/1132

File: src/libsyntax/attr.rs (hunk 11 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: Discussed with @nrc over IRC. Conclusion was that we'll keep the two traits since MetaItem and MetaListItems are logically different.

File: src/librustc_driver/lib.rs (hunk 20 above)
Comment (SergioBenitez) [inline src/librustc_driver/lib.rs]: The two different error messages should help out if the conditions are ever hit. I propose we keep them: doing so doesn't hurt and can only help.

File: src/libsyntax/ast.rs (hunk 21 above)
Comment (SergioBenitez) [inline src/libsyntax/ast.rs]: A `MetaItem` can already be nested in itself, but it cannot appear in a list. Thus, the ability to nest isn't the differentiation, it's the fact that the type can appear inside of a list. If others feels strongly about this, I'm okay with changing the name.

**Comment by SergioBenitez** [comment]
@nrc Would that be a problem though? That would only be in nightly, so it would be okay. Either someone is on stable and they don't get this change, or someone is only nightly/beta and they get this change and the macros.

**Comment by nikomatsakis** [comment]
I've not read the code but I trust that between @eddyb, @nrc, and @cgswords this is looking good. However, I agree with @nrc that we need a feature gate (and some tests). Many attributes are not as thorough as they should be about checking for random extra attributes and arguments and so forth, I suspect.

**Comment by nrc** [comment]
@SergioBenitez I'm not sure I understand what you're saying here - this feature will ride the trains, so in 12 weeks or so it will be on stable, let's say that I add an attribute today and forget to forbid use of lits, that gets stabilised quickly, and once it does (say in 24 weeks) it hits stable, then users on stable will be able to use this feature with that attribute.

File: src/librustc/hir/check_attr.rs [hunk 23]
```diff
@@ -53,34 +53,39 @@ impl<'a> CheckAttrVisitor<'a> {
...
-                    if target != Target::Struct {
-                        "attribute should be applied to struct"
-                    } else {
-                        continue
+                    "simd" => {
+                        if target != Target::Struct {
+                            "attribute should be applied to struct"
+                        } else {
+                            continue
+                        }
```
Comment (jseyfried) [inline src/librustc/hir/check_attr.rs]: nit: formatting (incorrect indent)

Comment (SergioBenitez) [inline src/librustc/hir/check_attr.rs]: This is also `rustfmt`. Doesn't look wrong to me though.

Comment (eddyb) [inline src/librustc/hir/check_attr.rs]: What it's doing is this:
```rust
                "packed" |
                    "simd" => {
                        if target != Target::Struct {
```
The arm body should not be indented twice from the arm itself.

Comment (SergioBenitez) [inline src/librustc/hir/check_attr.rs]: Ah, I see it now. Actually, this may have been my editor and not `rustfmt`.

Is it better to have ``"packed" | "simd"` on one line, or is it better the way it was before?

Comment (jseyfried) [inline src/librustc/hir/check_attr.rs]: The arm would look nicer (imo) and have less rightward drift as
```rust
                "packed" | "simd" => {
                    if target != Target::Struct {
                        "attribute should be applied to struct"
                    } else {
                        continue
                    }
                }
```
than
```rust
                "packed" |
                    "simd" => {
                        if target != Target::Struct {
                           "attribute should be applied to struct"
                        } else {
                            continue
                        }
                    }
```
I haven't seen the style with the extra indent anywhere in the compiler before, but I suppose `rustfmt` recently adopted it...

Regardless, orthogonal formatting changes make reviewing more difficult -- for future reference, perhaps do a separate commit for `rustfmt` changes.

Comment (nrc) [inline src/librustc/hir/check_attr.rs]: I'm not sure why this is happening - it makes sense if there are *a lot* of patterns and one line arm bodies, but not here. I think there is a combination of a bug (why we split at `|`) and a bad heuristic (double indent).

**Comment by SergioBenitez** [comment]
With the addition of the feature flag and a run-pass-fulldeps test, I believe I've addressed all of the comments. Is there anything else that needs to be done for this PR to land?

File: src/libsyntax/attr.rs [hunk 24]
```diff
@@ -357,6 +464,26 @@ pub fn last_meta_item_value_str_by_name(items: &[P<MetaItem>], name: &str)
...
+
+pub fn sort_meta_item(item: P<MetaItem>) -> P<MetaItem> {
+    item.map(|Spanned {node, span}| Spanned {
+        node: match node {
+            MetaItemKind::List(n, vs) => MetaItemKind::List(n, sort_meta_item_list_vals(vs)),
+            _ => node
+        },
+        span: span
+    })
+}
```
Comment (jseyfried) [inline src/libsyntax/attr.rs]: Both of these functions are just the identity (neither calls `sort_meta_items`), which does the real work.

n.b. `sort_meta_items` appears to be unused (both by `rustc` and by any plugins on Github) so it and these two functions you added can probably be removed.

File: src/libsyntax/attr.rs [hunk 25]
```diff
@@ -367,15 +494,7 @@ pub fn sort_meta_items(items: Vec<P<MetaItem>>) -> Vec<P<MetaItem>> {
...
-    v.into_iter().map(|(_, m)| m.map(|Spanned {node, span}| {
-        Spanned {
-            node: match node {
-                MetaItemKind::List(n, mis) => MetaItemKind::List(n, sort_meta_items(mis)),
-                _ => node
-            },
-            span: span
-        }
-    })).collect()
+    v.into_iter().map(|(_, m)| sort_meta_item(m)).collect()
```
Comment (jseyfried) [inline src/libsyntax/attr.rs]: This used to recursively sort `MetaItemKind::List`s but now does nothing.

Comment (jseyfried) [inline src/libsyntax/attr.rs]: (now that they can contain literals, sorting `MetaItemKind::List`s might not even make sense)

File: src/libsyntax/attr.rs (hunk 24 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: Good note. I'll remove them.

File: src/libsyntax/attr.rs (hunk 25 above)
Comment (SergioBenitez) [inline src/libsyntax/attr.rs]: Same as above.

File: src/librustc/hir/check_attr.rs (hunk 23 above)
Comment (nrc) [inline src/librustc/hir/check_attr.rs]: I'm not sure why this is happening - it makes sense if there are *a lot* of patterns and one line arm bodies, but not here. I think there is a combination of a bug (why we split at `|`) and a bad heuristic (double indent).

https://github.com/rust-lang-nursery/rustfmt/issues/1136

**Comment by SergioBenitez** [comment]
Rebased on master.

File: src/libsyntax/ast.rs (hunk 21 above)
Comment (nrc) [inline src/libsyntax/ast.rs]: I would like the name changes, MetaListItem doesn't seem very informative to me

**Comment by nrc** [comment]
r+ with MetaListItem renamed and with the commits squashed. Thanks for the PR, awesome work!

**Comment by SergioBenitez** [comment]
@nrc: Done.

**Comment by eddyb** [comment]
@Manishearth r=nrc

**Comment by SergioBenitez** [comment]
Rebased.

**Comment by jseyfried** [comment]
This will land in the next breaking batch.
@bors r-

**Comment by SergioBenitez** [comment]
@jseyfried Wasn't this supposed to be part of the previous batch? Don't see the harm in landing it now if so.

**Comment by jseyfried** [comment]
@SergioBenitez Which previous batch? Afaik, none of the pending syntax-[breaking-change] PRs have landed yet (c.f. https://github.com/rust-lang/rust/issues/31645#issuecomment-239228452).

**Comment by SergioBenitez** [comment]
@jseyfried: https://github.com/rust-lang/rust/pull/35979

**Comment by jseyfried** [comment]
Those aren't syntax-[breaking-change]s though.
